### PR TITLE
New LayeredConfig API

### PR DESCRIPTION
**Pull request by elandau**
LayeredConfig will eventually replace CompositeConfig as the top level container for properties.  LayeredConfig simplifies defining multiple override layers that may be added in any order while keeping a well-defined override order.  A set of common layers is defined in the Layers class but additional user defined layers may be defined and added at any time.

This PR is the first in the series of changes through which the LayeredConfig will eventually be the top level config API while keeping backwards compatibility in mind.  Once dependent libraries support a code path for LayeredConfig as the top level API, ArchaiusModule will be updated to use LayeredConfig instead of CompositeConfig.

### CODE REVIEW

File: archaius2-api/src/main/java/com/netflix/archaius/Layers.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import com.netflix.archaius.api.Layer;
+
+public final class Layers {
+    public static Layer TEST        = Layer.of("test",        100, false);
+    public static Layer RUNTIME     = Layer.of("runtime",     200, false);
+    public static Layer SYSTEM      = Layer.of("system",      300, false);
+    public static Layer ENVIRONMENT = Layer.of("environment", 400, false);
+    public static Layer REMOTE      = Layer.of("remote",      500, false);
+    public static Layer APPLICATION_OVERRIDE = Layer.of("application-override", 600, true);
```
Comment (twicksell) [inline archaius2-api/src/main/java/com/netflix/archaius/Layers.java]: Doesn't application-override typically beat remote?

File: archaius2-api/src/main/java/com/netflix/archaius/api/Layer.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * Key used to group and order {@link Config}s into layers (@see Layers).
+ * Layers are ordered by natural order such that lower order values have 
+ * precedence over a higher value.  Within a layer configurations are prioritized by
+ * insertion order (or reversed if 'reversed=true')
+ */
+public final class Layer {
+    private final String name;
+    private final int order;
+    private final boolean reversed;
```
Comment (twicksell) [inline archaius2-api/src/main/java/com/netflix/archaius/api/Layer.java]: Why do we need to keep this concept around? Is this truly necessary for a new API?

File: archaius2-api/src/main/java/com/netflix/archaius/api/config/LayeredConfig.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import com.netflix.archaius.api.Layer;
+import com.netflix.archaius.api.exceptions.ConfigException;
+
+import java.util.Collection;
+import java.util.Optional;
+
+/**
+ * Composite Config where the override order is driven by Layer keys.
+ */
+public interface LayeredConfig extends Config {
```
Comment (twicksell) [inline archaius2-api/src/main/java/com/netflix/archaius/api/config/LayeredConfig.java]: Long-term what happens to the Config interface? Would we even want to keep it? Will we continue to expose bindings for it? (beyond whatever is needed for backwards compatibility)

File: archaius2-api/src/main/java/com/netflix/archaius/api/Layer.java (hunk 2 above)
Comment (elandau) [reply archaius2-api/src/main/java/com/netflix/archaius/api/Layer.java]: I suppose we don't need this and can add it later if the need should arise.

File: archaius2-api/src/main/java/com/netflix/archaius/api/config/LayeredConfig.java (hunk 3 above)
Comment (elandau) [reply archaius2-api/src/main/java/com/netflix/archaius/api/config/LayeredConfig.java]: Yes, this is mostly for backwards compatibility.  Long term I'd like to get rid of Config in favor of a PropertyResolver and PropertySource, similar to what Spring has.  This should ease the migration in the future as well.  Once this has been released I'll be making more changes towards introducing PropertyResolver and PropertySource derived interfaces.

File: archaius2-api/src/main/java/com/netflix/archaius/Layers.java (hunk 1 above)
Comment (twicksell) [reply archaius2-api/src/main/java/com/netflix/archaius/Layers.java]: Nevermind, mixed up runtime and application-override in my head.

**Pull request merged**